### PR TITLE
feat: DAH-2541 directory statuses

### PR DESCRIPTION
**Pull request by tallulahkay**
## Description

Updates statuses on the image card of the directory page.

## Jira ticket

DAH-2541

### Version Control

- [x] branch name begins with `angular` if it contains updates to Angular code
- [x] branch name contains the Jira ticket number
- [x] PR name follows `type: TICKET-NUMBER Description` format, e.g. `feat: DAH-123 New Feature`

### Code quality

- [x] [the set of changes is small](https://google.github.io/eng-practices/review/developer/small-cls.html#what-is-small)
- [x] [if the set of changes cannot be small, reviewers have been forewarned](https://google.github.io/eng-practices/review/developer/small-cls.html#cant)
- [x] code meets test coverage thresholds
- [x] code is properly formatted
- [x] code is linted
- [x] code irrelevant to the ticket is not modified e.g. changing indentation due to automated formatting
- [x] automated tests pass
- [x] if the code changes the UI, it matches the UI design exactly

### Review instructions

- [x] instructions specify which environment(s) it applies to
- [x] instructions have already been performed at least once
- [x] instructions can be followed by PA testers
- [x] instructions specify if it can only be followed by an engineer

### Request review

- [x] PR has `needs review` label
- [x] Use `Housing Eng` group to automatically assign reviewers, and/or assign specific engineers
- [x] If time sensitive, notify engineers in Slack

### CODE REVIEW

**Comment by jimlin-sfgov** [comment]
question: For part 3 of the review instructions, I'm getting the red banner for this scenario
- searching
- `isFcfs = true`
- `isFcfsApplicationNotYetOpen = false`
- `isFcfsApplicationClosed = true`

Is this expected behavior?

Screenshot 2024-08-26 at 4 25 00 PM

**Comment by alulabeshue-sfgov** [comment]
thought: In `getListingImageCardStatuses` of SharedHelpers.tsx reorganizing the conditionals might make it a little more maintainable in the future if there are new listing types. Rather than having a condition based on closed or not and then nesting the fcfs logic into it we could have a condition based on listing type and nest the specific logic. Thoughts?

```
if (hasFiltersSet) {
  // return match or not match
} else if (isFcfs) {
  if (isFcfsApplicationClosed) {
    // return closed
  } else if (isFcfsApplicationNotYetOpen) {
    // return not yet open
  } else {
    // return open for application
  }
} else {
  // keep the same standard lottery logic as before the FCFS changes
}
```

File: app/assets/json/translations/react/en.json [hunk 1]
```diff
@@ -829,6 +829,12 @@
   "listingDetails.applicationsDeadline": "Applications Deadline",
   "listingDetails.applicationsDeadline.timezone": "Pacific Time",
   "listingDetails.applicationsOpen": "Applications Open",
+  "listingDirectory.listingStatusContent.applicationDeadline": "Application deadline",
+  "listingDirectory.listingStatusContent.applicationsClosed": "Applications closed",
+  "listingDirectory.listingStatusContent.applicationsOpen": "Applications open",
+  "listingDirectory.listingStatusContent.lotteryResultsPosted": "Lottery results posted",
+  "listingDirectory.listingStatusContent.subContent.firstComeFirstServed": "First come, first served",
+  "listingDirectory.lotteryResultsPosted": "",
```
Comment (alulabeshue-sfgov) [inline app/assets/json/translations/react/en.json:837]: issue: With these changes the image card status is not translated even for non FCFS listings. Should we continue to use the translated content until we get translations back?

Comment (tallulahkay) [reply app/assets/json/translations/react/en.json:837]: Good call!! Maybe I should just put this change behind a feature flag?

**Comment by tallulahkay** [comment]
> question: For part 3 of the review instructions, I'm getting the red banner for this scenario
> 
> * searching
> * `isFcfs = true`
> * `isFcfsApplicationNotYetOpen = false`
> * `isFcfsApplicationClosed = true`
> 
> Is this expected behavior?
> 
> Screenshot 2024-08-26 at 4 25 00 PM

Thanks for testing that Jim! I think that is actually the full environment not being appropriate for testing (and my review plan not being thorough enough) - in reality, if that listing is closed, it shouldn't show up as a match. I'll think on how to test this better

**Comment by tallulahkay** [comment]
> thought: In `getListingImageCardStatuses` of SharedHelpers.tsx reorganizing the conditionals might make it a little more maintainable in the future if there are new listing types. Rather than having a condition based on closed or not and then nesting the fcfs logic into it we could have a condition based on listing type and nest the specific logic. Thoughts?
> 
> ```
> if (hasFiltersSet) {
>   // return match or not match
> } else if (isFcfs) {
>   if (isFcfsApplicationClosed) {
>     // return closed
>   } else if (isFcfsApplicationNotYetOpen) {
>     // return not yet open
>   } else {
>     // return open for application
>   }
> } else {
>   // keep the same standard lottery logic as before the FCFS changes
> }
> ```

Yeah you're so right - we can even do separate functions for getting the statuses depending on the listing type

File: app/assets/json/translations/react/en.json (hunk 1 above)
Comment (alulabeshue-sfgov) [reply app/assets/json/translations/react/en.json:837]: Yea makes sense to me. I saw in [DAH-2588](https://github.com/SFDigitalServices/sf-dahlia-web/pull/2240/files) that we created a `see_the_unit` flag. Im ok with having a separate FCFS flag that way the see the unit feature can be released independently, once we get back translations for the content, without having to wait for the entire FCFS feature.

**Review by alulabeshue-sfgov — COMMENTED** [review]
> > thought: In `getListingImageCardStatuses` of SharedHelpers.tsx reorganizing the conditionals might make it a little more maintainable in the future if there are new listing types. Rather than having a condition based on closed or not and then nesting the fcfs logic into it we could have a condition based on listing type and nest the specific logic. Thoughts?
> > ```
> > if (hasFiltersSet) {
> >   // return match or not match
> > } else if (isFcfs) {
> >   if (isFcfsApplicationClosed) {
> >     // return closed
> >   } else if (isFcfsApplicationNotYetOpen) {
> >     // return not yet open
> >   } else {
> >     // return open for application
> >   }
> > } else {
> >   // keep the same standard lottery logic as before the FCFS changes
> > }
> > ```
> 
> Yeah you're so right - we can even do separate functions for getting the statuses depending on the listing type

Adding this as a review to have start a thread...

Yea. I was thinking if we can create a util function that tells us if a listing is FCFS and what state it is in (not yet open, open, closed) then we can clean up some duplicate logic in SharedHelpers, ListingDetailsApply, ListingDetailsApplicationDate, and any other components that will need this logic.

***EDIT:*** I did something wrong, did not start a thread :(

**Comment by tallulahkay** [comment]
@alulabeshue-sfgov @jimlin-sfgov do not know how to thread - but take a look I did a little refactoring

**Review by alulabeshue-sfgov — APPROVED** [review]
LG! 

Thanks for all of the clean up, looking really good! I think we can skip the code climate test checks for now. Im working on getting the fields we need back from salesforce for FCFS status which will make testing SharedHelpers easier. All of the untested code is behind a feature flag for now.

**Pull request merged**